### PR TITLE
[Plugin RpcServer] fix issue 3629

### PR DESCRIPTION
**Pull request by Jim8y**
# Description

This pr fixes https://github.com/neo-project/neo/issues/3629 to handle rpcexception catch.

Fixes # https://github.com/neo-project/neo/issues/3629

## Type of change

- [ ] Optimization (the change is only an optimization)
- [ ] Style (the change is only a code style for better maintenance or standard purpose)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] TestInvokeFunctionInvalid

**Test Configuration**:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/Plugins/RpcServer/RpcServer.cs [hunk 1]
```diff
@@ -333,7 +335,8 @@ private async Task<JObject> ProcessRequestAsync(HttpContext context, JObject req
                             {
                                 args[i] = param.DefaultValue;
                             }
-                            else if (param.ParameterType.IsValueType && Nullable.GetUnderlyingType(param.ParameterType) == null)
+                            else if (param.ParameterType.IsValueType &&
```
Comment (Jim8y) [inline src/Plugins/RpcServer/RpcServer.cs:338]: these are made automatically by dotnet format.

File: src/Plugins/RpcServer/RpcServer.cs [hunk 2]
```diff
@@ -366,11 +369,16 @@ private async Task<JObject> ProcessRequestAsync(HttpContext context, JObject req
...
+                return CreateErrorResponse(request["id"],
+                    RpcErrorFactory.NewCustomError(ex.HResult, ex.Message, ex.StackTrace));
 #else
         return CreateErrorResponse(request["id"], RpcErrorFactory.NewCustomError(ex.HResult, ex.Message));
 #endif
             }
+            catch (RpcException ex)
+            {
+                return CreateErrorResponse(request["id"], ex.GetError());
+            }
```
Comment (Jim8y) [inline src/Plugins/RpcServer/RpcServer.cs:381]: core update is here

**Review by Hecate2 — DISMISSED** [review]
May be better if `ParameterConverter.ConvertParameter` throws something else, instead of RpcException.

**Review by roman-khimov — COMMENTED** [review]
What's gonna be the HTTP return code this way? 200?

**Comment by Hecate2** [comment]
> What's gonna be the HTTP return code this way? 200?

500 with no response body

**Comment by Hecate2** [comment]
> What's gonna be the HTTP return code this way? 200?

200 with an error code like -32602 in the json response body

**Review by superboyiii — APPROVED** [review]
It works correctly, although it changes the original behaviour for message. I'm not sure it's good or not since these messages are not useless for debug.
Original(v3.7.5):
38cab1448ae3a80926cbe334a897253
Current:
c9912bc94db6e7d3122e2975745065d

File: src/Plugins/RpcServer/RpcServer.cs [hunk 3]
```diff
@@ -366,11 +369,16 @@ private async Task<JObject> ProcessRequestAsync(HttpContext context, JObject req
...
-                return CreateErrorResponse(request["id"], RpcErrorFactory.NewCustomError(ex.HResult, ex.Message, ex.StackTrace));
+                return CreateErrorResponse(request["id"],
+                    RpcErrorFactory.NewCustomError(ex.HResult, ex.Message, ex.StackTrace));
 #else
         return CreateErrorResponse(request["id"], RpcErrorFactory.NewCustomError(ex.HResult, ex.Message));
 #endif
             }
+            catch (RpcException ex)
+            {
+                return CreateErrorResponse(request["id"], ex.GetError());
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply src/Plugins/RpcServer/RpcServer.cs:380]: Or at least we may use `#if DEBUG` directive to provide extended error message with trace.

[the remark this answers is not in the document]
Comment (Jim8y) [reply src/Plugins/RpcServer/RpcServer.cs:380]: will add

[the remark this answers is not in the document]
Comment (Jim8y) [reply src/Plugins/RpcServer/RpcServer.cs:385]: Applied

**Review by Hecate2 — APPROVED** [review]

**Review by shargon — APPROVED** [review]

**Review by AnnaShaleva — APPROVED** [review]

**Pull request merged**